File: examples/kotlin/nats-client/src/main/kotlin/cdcexamples/nats/ChangesClient.kt
package cdcexamples.nats

import java.net.URI
import java.net.http.HttpClient
import java.net.http.HttpRequest
import java.net.http.HttpResponse
import java.time.Duration

/**
 * ChangesClient holt die Änderungen eines Quelle/Schema/Tabelle-Filters mit
 * dem `reader`-Token ab (`LH-FA-SST-006`, zweiseitiger Ablauf `ADR-0079`) —
 * Form-Vorbild: `examples/nats-client/main.go` (`fetchChanges`),
 * `examples/kotlin/http-client/TablesClient.kt`. Ein Nicht-200-Status ist
 * ein sichtbarer Fehler, kein stiller Leerwert. Der [HttpClient] ist
 * injiziert, damit der Fehlerpfad ohne erreichbaren Host netzlos testbar ist.
 */
object ChangesClient {
    private val requestTimeout: Duration = Duration.ofSeconds(10)

    fun fetch(httpClient: HttpClient, cfg: Config): String {
        val url = ChangesUrl.build(cfg.addr, cfg.source, cfg.schema, cfg.table)
        val request = HttpRequest.newBuilder(URI.create(url))
            .header("Authorization", "Bearer ${cfg.token}")
            .timeout(requestTimeout)
            .GET()
            .build()

        val response = httpClient.send(request, HttpResponse.BodyHandlers.ofString())
        if (response.statusCode() != 200) {
            throw IllegalStateException("HTTP-Status ${response.statusCode()}: ${response.body()}")
        }
        return response.body()
    }
}

File: examples/csharp/nats-client/ChangesClient.cs
using System.Net;
using System.Net.Http.Headers;

namespace CdcExamples.Nats;

/// <summary>
/// ChangesClient holt die Änderungen eines Quelle/Schema/Tabelle-Filters mit
/// dem <c>reader</c>-Token ab (<c>LH-FA-SST-006</c>, zweiseitiger Ablauf
/// <c>ADR-0079</c>) — Form-Vorbild: <c>examples/nats-client/main.go</c>
/// (<c>fetchChanges</c>). Ein Nicht-200-Status ist ein sichtbarer Fehler,
/// kein stiller Leerwert. Der <see cref="System.Net.Http.HttpClient"/> ist
/// injiziert, damit der Fehlerpfad ohne erreichbaren Host netzlos testbar ist.
/// </summary>
public static class ChangesClient
{
    public static async Task<string> FetchAsync(
        System.Net.Http.HttpClient httpClient,
        Config cfg,
        CancellationToken cancellationToken = default)
    {
        var url = ChangesUrlBuilder.Build(cfg.Addr, cfg.Source, cfg.Schema, cfg.Table);
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.Authorization = new AuthenticationHeaderValue("Bearer", cfg.Token);

        using var response = await httpClient.SendAsync(request, cancellationToken).ConfigureAwait(false);
        var body = await response.Content.ReadAsStringAsync(cancellationToken).ConfigureAwait(false);
        if (response.StatusCode != HttpStatusCode.OK)
        {
            throw new InvalidOperationException($"HTTP-Status {(int)response.StatusCode}: {body}");
        }
        return body;
    }
}
